app/Listeners/Network/RecordFirEntry.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+namespace App\Listeners\Network;
+
+use App\Events\NetworkAircraftUpdatedEvent;
+use App\Models\FlightInformationRegion\FlightInformationRegion;
+use App\Models\Vatsim\NetworkAircraftFirEvent;
+use App\Services\FlightInformationRegionService;
+use Location\Coordinate;
+use Location\Polygon;
+
+class RecordFirEntry
+{
+    /**
+     * @var Polygon[]
+     */
+    private $boundaries;
+
+    public function __construct()
+    {
+        $this->boundaries = FlightInformationRegion::all()
+            ->mapWithKeys(function (FlightInformationRegion $region) {
+                return [$region->id => FlightInformationRegionService::getBoundaryPolygon($region->identification_code)];
+            })
+            ->toArray();
+    }
+
+    public function handle(NetworkAircraftUpdatedEvent $event): bool
+    {
+        $lastEntry = NetworkAircraftFirEvent::where('callsign', $event->getAircraft()->callsign)
+            ->where('event_type', 'FIR_ENTRY')
+            ->latest('created_at')
+            ->first();
+
+        // Handle FIR exit
+        if (
+            $lastEntry &&
+            !$this->hasPreviouslyExitedFir($event->getAircraft()->callsign, $lastEntry->flight_information_region_id) &&
+            isset($this->boundaries[$lastEntry->flight_information_region_id]) &&
+            !$this->boundaries[$lastEntry->flight_information_region_id]->contains(new Coordinate($event->getAircraft()->latitude,
+                $event->getAircraft()->longitude))
+        ) {
+            NetworkAircraftFirEvent::create(
+                [
+                    'callsign' => $event->getAircraft()->callsign,
+                    'flight_information_region_id' => $lastEntry->flight_information_region_id,
+                    'event_type' => 'FIR_EXIT',
+                    'metadata' => [
+                        'exit_latitude' => $event->getAircraft()->latitude,
+                        'exit_longitude' => $event->getAircraft()->longitude,
+                    ],
+                ]
+            );
+        }
+
+        // Handle FIR entry
+        foreach ($this->boundaries as $firId => $boundary) {
+
+            // Don't record entering the same FIR unless it has since exited.
+            if ($lastEntry && $firId === $lastEntry->flight_information_region_id) {
+                $hasExited = NetworkAircraftFirEvent::where('callsign', $event->getAircraft()->callsign)
+                    ->where('flight_information_region_id', $lastEntry->flight_information_region_id)
+                    ->where('event_type', 'FIR_EXIT')
```

Comment:
Is it worth defining `FIR_EXIT` and `FIR_ENTRY` as a const on the model?

Reply:
Done!